qiita_db/handlers/processing_job.py

```diff
@@ -0,0 +1,157 @@
+            if success:
+                job_status = job.status
+                if job_status == 'queued':
+                    job.status = 'running'
+                    job.heartbeat = datetime.now()
+                elif job_status == 'running':
+                    job.heartbeat = datetime.now()
+                else:
+                    success = False
+                    error_msg = 'Job already finished. Status: %s' % job_status
+
+        response = {'success': success, 'error': error_msg}
+        self.write(response)
+
+
+class StepHandler(RequestHandler):
+    def post(self, job_id):
+        """Changes the step of the given job
+
+        Parameters
+        ----------
+        job_id : str
+            The job to change the step
+        """
+        with qdb.sql_connection.TRN:
+            job, success, error_msg = _get_job(job_id)
+            if success:
+                job_status = job.status
+                if job_status != 'running':
+                    success = False
+                    error_msg = 'Job in a non-running state'
+                else:
+                    payload = loads(self.request.body)
+                    step = payload['step']
+                    job.step = step
+
+        response = {'success': success, 'error': error_msg}
+        self.write(response)
+
+
+class CompleteHandler(RequestHandler):
+    def post(self, job_id):
+        """Updates the job to one of the completed status ('success', 'error')
+
+        Parameters
+        ----------
+        job_id : str
+            The job to complete
+        """
+        with qdb.sql_connection.TRN:
+            job, success, error_msg = _get_job(job_id)
+            if success:
+                if job.status != 'running':
+                    success = False
+                    error_msg = 'Job in a non-running state.'
+                else:
+                    payload = loads(self.request.body)
+                    if payload['success']:
+                        for artifact_data in payload['artifacts']:
+                            filepaths = artifact_data['filepaths']
+                            atype = artifact_data['artifact_type']
+                            parents = job.input_artifacts
+                            params = job.parameters
+                            ebi = artifact_data['can_be_submitted_to_ebi']
+                            vamps = artifact_data['can_be_submitted_to_vamps']
+                            qdb.artifact.Artifact.create(
+                                filepaths, atype, parents=parents,
+                                processing_parameters=params,
+                                can_be_submitted_to_ebi=ebi,
+                                can_be_submitted_to_vamps=vamps)
+                        job.status = 'success'
+                    else:
+                        log = qdb.logger.LogEntry.create(
+                            'Runtime', payload['error'])
```

Comment:
Not sure about this, but the paylod will include the information about why something failed, correct?